elastic/kibana
# [TSVB] add rison helper and URL encoding for drilldown urls

## Summary
Resolves https://github.com/elastic/kibana/issues/24233 by
  - removing the default handlebars HTML escaping behavior for drilldown URLs
  - introducing a `rison` and an `encodeURIComponent` custom helper

## To Test
### Setup
#### Create a timeseries index with some documents

```
PUT /test_timeseries
{
  "mappings": {
    "properties" : {
      "@timestamp" : {
        "type" : "date"
      },
      "foo" : {
        "type" : "text",
        "fields" : {
          "keyword" : {
            "type" : "keyword",
            "ignore_above" : 256
          }
        }
      }
    }
  }
}

POST /test_timeseries/_doc
{
  "@timestamp": 1643661888326,
  "foo": "!bar"
}

POST /test_timeseries/_doc
{
  "@timestamp": 1643661888326,
  "foo": "#bar"
}

POST /test_timeseries/_doc
{
  "@timestamp": 1643661888326,
  "foo": "&bar"
}
```

#### Import the [saved objects](https://drive.google.com/file/d/1qRPw25k91aYm0HetKojkI3veqdSDQwx6/view?usp=sharing)

### Manual test
- Visit the imported dashboard called "Drilldown Origin Dash"
- Adjust time range to last 1 year
- Click cog -> "Edit visualization" on the table you see
- Click "Panel options"

#### Works for invalid URL characters ("#", "&")
- for "Item url," enter the "With URL Encoding" URL you see below
- click "#bar" in the table
- saved search should load without errors and only the document with "#bar" is displayed
- click "&bar" in the table
- saved search should load without errors and only the document with "&bar" is displayed
- click "!bar" in the table
- saved search shows errors

#### Works for invalid Rison character ("!")
- for "Item url," enter the "With RISON" URL you see below
- click "!bar" in the table
- saved search should load without errors and only the document with "!bar" is displayed
- click "&bar" or "#bar" in the table
- saved search shows errors

#### Works for everything
- for "Item url," enter the "With BOTH" URL you see below
- for each entry in the table
  - click
  - saved search should load without errors and only the document with that entry is displayed

#### Test URLS
```
With RISON
<BASE_KIBANA_PATH>/app/discover#/view/0ac50180-82d9-11ec-9f4a-55de56b00cc0?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15d,to:now))&_a=(columns:!(),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:d29125b0-82d7-11ec-9f4a-55de56b00cc0,key:foo.keyword,negate:!f,params:(query:%23bar),type:phrase),query:(match_phrase:(foo.keyword:{{rison key}})))),grid:(),hideChart:!f,index:d29125b0-82d7-11ec-9f4a-55de56b00cc0,interval:auto,query:(language:kuery,query:''),sort:!(!('@timestamp',desc)))

With URL Encoding
<BASE_KIBANA_PATH>/app/discover#/view/0ac50180-82d9-11ec-9f4a-55de56b00cc0?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15d,to:now))&_a=(columns:!(),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:d29125b0-82d7-11ec-9f4a-55de56b00cc0,key:foo.keyword,negate:!f,params:(query:%23bar),type:phrase),query:(match_phrase:(foo.keyword:{{encodeURIComponent key}})))),grid:(),hideChart:!f,index:d29125b0-82d7-11ec-9f4a-55de56b00cc0,interval:auto,query:(language:kuery,query:''),sort:!(!('@timestamp',desc)))

With BOTH
<BASE_KIBANA_PATH>/app/discover#/view/0ac50180-82d9-11ec-9f4a-55de56b00cc0?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15d,to:now))&_a=(columns:!(),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:d29125b0-82d7-11ec-9f4a-55de56b00cc0,key:foo.keyword,negate:!f,params:(query:%23bar),type:phrase),query:(match_phrase:(foo.keyword:{{encodeURIComponent (rison key)}})))),grid:(),hideChart:!f,index:d29125b0-82d7-11ec-9f4a-55de56b00cc0,interval:auto,query:(language:kuery,query:''),sort:!(!('@timestamp',desc)))
```
### Checklist

Delete any items that are not applicable to this PR.
